clusterprovider/kcp/kcp.go

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providerkcp
+
+import (
+	"context"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+
+	k8sapierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/clientcmd"
+	api "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog/v2"
+
+	corev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	tenancyclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/typed/tenancy/v1alpha1"
+
+	clusterprovider "github.com/kubestellar/kubestellar/pkg/clustermanager/providerclient"
+)
+
+// KcpClusterProvider is a cluster provider that works with a local Kind instance.
+type KcpClusterProvider struct {
+	ctx        context.Context
+	logger     logr.Logger
+	configPath string
+	clientset  tenancyclient.WorkspaceInterface
+	workspaces map[string]string
```

Comment:
It's a pretty nice to have the list of workspaces at the kcp provider level.  But do we actually make use of it other than setting and deleting from it?

Reply:
Correct, the value is not used. I add to this cache map only when workspace is ready and KCP provider sent an event about it. I than check that workspace is cached so the event will not be duplicated.